### PR TITLE
New division implementation based on uint64 words

### PR DESCRIPTION
**Pull request by chfast**
1. This mostly replaces 32-bit word based Knuth division with 64-bit based Knuth division.
2. Implementation is ported from https://github.com/chfast/intx : https://github.com/chfast/intx/blob/33580022735f9c0595168c98656ab9b3b3fa16df/lib/intx/div.cpp.
3. The "quotient one too big" adjustments are not implemented. They are not needed for correctness but can improve performance. Alternatively, division 3by2 can be used in place.
4. We are officially `big.Int` free.
5. There are still dynamic allocations needed. This to be addressed later.

```
name                     old time/op  new time/op  delta
_Div/large/big-6          247ns ± 0%   245ns ± 0%   -0.81%  (p=0.008 n=5+5)
_Div/large/uint256-6      185ns ± 0%   139ns ± 0%  -24.86%  (p=0.000 n=5+4)
_Div/small/big-6         68.8ns ± 0%  68.7ns ± 0%     ~     (p=0.714 n=4+5)
_Div/small/uint256-6     13.2ns ± 0%  13.0ns ± 0%   -1.52%  (p=0.008 n=5+5)
_MulMod/large/big-6       386ns ± 0%   388ns ± 0%   +0.41%  (p=0.016 n=5+4)
_MulMod/large/uint256-6   535ns ± 0%   245ns ± 0%  -54.21%  (p=0.008 n=5+5)
_MulMod/small/big-6      94.4ns ± 0%  93.8ns ± 0%   -0.64%  (p=0.008 n=5+5)
_MulMod/small/uint256-6  41.2ns ± 0%  40.4ns ± 0%   -2.04%  (p=0.008 n=5+5)
_Mod/large/big-6          137ns ± 0%   136ns ± 0%   -0.73%  (p=0.029 n=4+4)
_Mod/large/uint256-6      149ns ± 0%    81ns ± 0%  -45.89%  (p=0.008 n=5+5)
_Mod/small/big-6         41.2ns ± 0%  40.7ns ± 0%   -1.12%  (p=0.000 n=5+4)
_Mod/small/uint256-6     14.0ns ± 0%  13.9ns ± 0%   -0.71%  (p=0.008 n=5+5)
_SDiv/large/big-6         405ns ± 0%   401ns ± 0%   -0.99%  (p=0.008 n=5+5)
_SDiv/large/uint256-6     193ns ± 0%   146ns ± 0%  -24.35%  (p=0.008 n=5+5)
_AddMod/large/big-6       212ns ± 0%   212ns ± 0%     ~     (all equal)
_AddMod/large/uint256-6  39.5ns ± 0%  39.6ns ± 0%     ~     (p=0.167 n=5+5)
_AddMod/small/big-6      66.2ns ± 0%  66.1ns ± 1%     ~     (p=0.143 n=4+5)
_AddMod/small/uint256-6  17.4ns ± 0%  17.4ns ± 0%     ~     (all equal)
```

### CODE REVIEW

File: uint256.go [hunk 1]
```diff
@@ -369,100 +368,124 @@ func (z *Int) isBitSet(n uint) bool {
 	return (z[n>>6] & (1 << (n & 0x3f))) != 0
 }
 
-func nlz(d *Int) uint {
-	for i := 3; i >= 0; i-- {
-		if d[i] != 0 {
-			return uint(bits.LeadingZeros64(d[i]) % 32)
-		}
+func addTo(x, y []uint64) uint64 {
```
Comment (holiman) [inline uint256.go]: Even though it's internal, can we have a bit of method doc here?

File: uint256.go [hunk 2]
```diff
@@ -369,100 +368,124 @@ func (z *Int) isBitSet(n uint) bool {
...
-	var prev, t uint32
-	for i := 0; i <= n; i++ {
-		t = dn[i]
-		dn[i] = prev | (dn[i] << s)
-		prev = t >> (32 - s)
-	}
-	if isdividend {
-		n = n + 1
-		dn[n] = prev
+func subMulTo(x, y []uint64, multiplier uint64) uint64 {
```
Comment (holiman) [inline uint256.go]: Here too, would be good with at least a bit of doc

File: uint256.go [hunk 3]
```diff
@@ -369,100 +368,124 @@ func (z *Int) isBitSet(n uint) bool {
...
-				k = t >> 32
-			}
-			x[j+n] = x[j+n] + uint32(k)
+
+		quot[j] = qhat // Store quotient digit.
+	}
+	return quot
+}
+
+func udivrem(u []uint64, d *Int) (quot []uint64, rem *Int) {
```
Comment (holiman) [inline uint256.go]: This is still implementing `// See Knuth, Volume 2, section 4.3.1, Algorithm D.`, right? please make a docstring

**Comment by holiman** [comment]
Sweet -- once this is in, I believe that this lib is faster than big.Int in every single op -- and in most places by a very wide margin!

**Comment by chfast** [comment]
Added documentation. Fixup commits to be squashed before merge.

**Pull request merged**